Clarity Wasm

hugoclrd opened the pull request:
This contains the changes used for clarity-wasm. It may not yet be ready to be merged, but the clarity-wasm team is reviewing it as we go and then it will need a thorough review once it is ready to be merged into next.

This branch supersedes https://github.com/stacks-network/stacks-core/pull/3880, https://github.com/stacks-network/stacks-core/pull/4093, and the previous branch that was used by clarinet. The clarinet specific changes (to make the clarity crate compilable to wasm) are the starting point for the clarity-wasm changes.

hugocaillard commented on clarity/src/vm/database/sqlite.rs:
```diff
@@ -73,6 +82,62 @@ fn sqlite_has_entry(conn: &Connection, key: &str) -> Result<bool> {
+    Ok((bhh, contract_hash))
+}
+
+pub fn sqlite_insert_metadata(
+    store: &mut dyn ClarityBackingStore,
+    contract: &QualifiedContractIdentifier,
+    key: &str,
+    value: &str,
+) -> Result<()> {
+    let bhh = store.get_open_chain_tip();
+    SqliteConnection::insert_metadata(
+        store.get_side_store(),
+        &bhh,
+        &contract.to_string(),
+        key,
+        value,
+    )
+}
+
+pub fn sqlite_get_metadata(
+    store: &mut dyn ClarityBackingStore,
+    contract: &QualifiedContractIdentifier,
+    key: &str,
+) -> Result<Option<String>> {
+    let (bhh, _) = store.get_contract_hash(contract)?;
+    SqliteConnection::get_metadata(store.get_side_store(), &bhh, &contract.to_string(), key)
+}
+
+pub fn sqlite_get_metadata_manual(
+    store: &mut dyn ClarityBackingStore,
+    at_height: u32,
+    contract: &QualifiedContractIdentifier,
+    key: &str,
+) -> Result<Option<String>> {
+    let bhh = store.get_block_at_height(at_height).ok_or_else(|| {
+        warn!("Unknown block height when manually querying metadata"; "block_height" => at_height);
+        RuntimeErrorType::BadBlockHeight(at_height.to_string())
+    })?;
+    SqliteConnection::get_metadata(store.get_side_store(), &bhh, &contract.to_string(), key)
+}
```
These functions seems to be unused / irrelevant

hugocaillard commented on clarity/src/vm/types/signatures.rs:
```diff
@@ -873,6 +873,48 @@ impl TypeSignature {
+                },
+            ))),
+            TypeSignature::TupleType(TupleTypeSignature { type_map }) => {
+                let mut type_map_ref = type_map.as_ref().clone();
```
The `type_map` type change from `HashMap<>` to `Arc<BtreeMap>` 
I'm not sure about this `as_ref().clone()` that I had to do

obycode commented:
Let's get any failing tests working again in this branch.

matteojug committed:
fix: add all of the required duplicated code

matteojug committed:
Merge pull request #5135 from stacks-network/fix/wasm-print-stub

Clarity-Wasm: fix for the print host function

matteojug committed:
chore: avoid using testing-gated parse function

matteojug committed:
Merge pull request #5167 from stacks-network/chore/use-nontesting-parse-fn

Clarity-Wasm: avoid using testing-gated parse function

matteojug committed:
feat: add error mapping for already used arg throw

matteojug committed:
chore: fmt

matteojug committed:
Merge pull request #5192 from stacks-network/fix/throw-already-used-arg-error

Clarity-Wasm: add error mapping for already used arg error

Commit:
Merge branch 'develop' into chore/clarity-wasm-merge-develop

csgui committed:
Merge pull request #5239 from stacks-network/chore/clarity-wasm-merge-develop

clarity-wasm merge develop

obycode committed:
Merge branch 'develop' into feat/clarity-wasm-develop

Commit:
chore: sync host functions with clarity-wasm implementation

Commit:
fix: remove wsts library

csgui committed:
Merge pull request #5293 from stacks-network/sync_wasm_host_functions

chore: sync host functions with clarity-wasm implementation

hugoclrd committed:
fix: don not use rusqlite in clarity-vm without default features

Commit:
Add stub for `debug_msg` in linker

Commit:
spaces instead of tab

krl committed:
Merge pull request #5298 from stacks-network/debug_msg

Add stub for `debug_msg` in linker

Commit:
fix: do not panic! on ErrorMap::Panic

csgui committed:
Merge pull request #5306 from stacks-network/fix-error-map

fix ErrorMap

Commit:
fix: correct error mapping for BadTypeConstruction

Commit:
chore: document ErrorMap

csgui committed:
Merge pull request #5309 from stacks-network/fix-error-map

fix: correct error mapping for BadTypeConstruction

Commit:
Merge branch 'develop' into update-clarity-wasm-develop

csgui committed:
Merge pull request #5316 from stacks-network/update-clarity-wasm-develop

Update clarity wasm develop

Commit:
chore: sync with clarity-wasm repo

Commit:
chore: fmt

Commit:
fix: add missing dependencies

Commit:
chore: fmt

Commit:
fix: add missing dependency

csgui committed:
Merge pull request #5325 from stacks-network/update-clarity-wasm-develop

Update clarity wasm develop

Commit:
Merge branch 'develop' into update-clarity-wasm-develop

csgui committed:
Merge pull request #5331 from stacks-network/update-clarity-wasm-develop

Update clarity wasm develop

Commit:
chore: update Cargo.lock

csgui committed:
Merge pull request #5339 from stacks-network/update-clarity-wasm-develop

Update Cargo.lock on feat/clarity-wasm-develop

Commit:
chore: fix rusqlite dependency path

Commit:
chore: fmt

csgui committed:
Merge pull request #5340 from stacks-network/update-clarity-wasm-develop

fix rusqlite dependency path

Commit:
Merge tag '3.0.0.0.0' into clarity-wasm-3-0-0-0

Commit:
chore: update Cargo.lock

csgui committed:
Merge pull request #5373 from stacks-network/clarity-wasm-3-0-0-0

Update ClarityWasm with 3.0.0.0.0

Commit:
chore: update linnker functions

Commit:
chore: sync wasm changes for argument errors

Commit:
chore: update Cargo.lock

Commit:
chore: fmt

Commit:
fix: dependencies

csgui committed:
Merge pull request #5384 from stacks-network/update-linker-functions

update linker functions

Commit:
chore: update clarity-wasm

Commit:
chore: update clarity-wasm

Commit:
chore: fmt

csgui committed:
Merge pull request #5401 from stacks-network/update-clarity-wasm

Update feat/clarity-wasm-develop branch

ureeves committed:
feat(wasm): use only one `Engine` per global context

Instantiating a `wasmtime::Engine` is an expensive operation, so it is
best to do it only once for the duration of a `GlobalContext`. Cloning
an engine is, however, not an expensive operation (just an `Arc::clone`)
and we use it to avoid referring to the global context already referred
to by the instantiated `ClarityWasmContext`.

See-also: https://github.com/stacks-network/clarity-wasm/issues/468

obycode committed:
Merge pull request #5447 from ureeves/single-wasmtime-engine

feat(wasm): use only one `Engine` per global context

hugoclrd committed:
refactor: add serde default to symbolic expression struct

BowTiedWoo committed:
feat: split get-(burn)-block-info linker function

hugoclrd committed:
fix: add clarity-wasm feature flag in context

hugoclrd committed:
refactor: address review

BowTiedWoo committed:
fix: pass `Memory` as value

Commit:
chore: update clar2wasm dependency

csgui committed:
Merge pull request #5503 from stacks-network/chore/update-clar2wasm

update clar2wasm dependency

BowTiedWoo committed:
feat: wrap in `Optional` return type of validate height function

BowTiedWoo committed:
feat: add linker fn for `get-tenure-info?` and `get-stacks-block-info?`

hugoclrd committed:
Merge branch 'develop' into clarity-wasm-merge-develop

Commit:
chore: add missing functions to ClarityBackingStore trait

Commit:
chore: add missing dependency

Commit:
chore: add missing trait functions implementations

Commit:
chore: add missing trait functions implementations

Commit:
chore: add missing dependency

Commit:
chore: fmt

csgui committed:
Merge pull request #5512 from stacks-network/chore/add_missing_functions_to_trait

add missing functions to trait

hugoclrd committed:
Merge pull request #5509 from stacks-network/clarity-wasm-merge-develop

Merge develop onto clarity wasm

Commit:
chore: update clarity-wasm dependency

Acaccia committed:
Merge pull request #5481 from BowTiedWoo/feat/clarity-wasm-split-get-(burn)-block-info-linker-fn

Split `get-(burn)-block-info` linker function

BowTiedWoo committed:
Merge branch 'feat/clarity-wasm-develop' into feat/clarity-wasm-support-clarity-3-functions

BowTiedWoo committed:
fix: remove unused linker functions

BowTiedWoo committed:
Merge pull request #5506 from stacks-network/feat/clarity-wasm-support-clarity-3-functions

Add linker functions for `get-tenure-info?` and `get-stacks-block-info?`

hugoclrd committed:
feat: add serialize to cost synthesis

hugoclrd committed:
Merge pull request #5534 from stacks-network/feat/serialize-cost-synthesis

feat: add serialize to cost synthesis

hugoclrd committed:
Merge branch 'develop' into chore/merge-develop

hugoclrd committed:
Merge pull request #5580 from stacks-network/chore/merge-develop

Chore/merge develop

Commit:
chore: update Cargo.lock

csgui committed:
Merge pull request #5585 from stacks-network/sync-clarity-wasm-main

update Cargo.lock

Commit:
Merge branch 'develop' into clarity-wasm-merge-develop

Commit:
chore: cargo fmt

Commit:
chore: add wasm as a Cargo feature

Commit:
test: fix param mutability

csgui committed:
Merge pull request #5690 from stacks-network/clarity-wasm-merge-develop

merge develop into feat/clarity-wasm-develop

Commit:
Merge branch 'develop' into develop-into-clarity-wasm-develop

Commit:
chore: remove unused imports

Commit:
fix: adapt code to refactor on StacksAddress and PrincipalData

Commit:
chore: fix clarity-wasm dependency branch

Commit:
chore: fmt

csgui committed:
Merge pull request #5852 from stacks-network/develop-into-clarity-wasm-develop

Merge develop into feat/clarity-wasm-develop

hugoclrd committed:
fix: remove unwanted compilation flag

aldur committed:
Merge pull request #5878 from hugocaillard/fix/clarity-wasm-develop-compilation

fix: remove unwanted compilation flag

Commit:
Merge branch 'develop' into feat/clarity-wasm-develop

Commit:
chore: sync with develop and update Cargo.lock

Commit:
feat: add generic expect wasm error

csgui committed:
Merge pull request #5906 from stacks-network/feat/clarity-wasm-generic-expect-error

wasm generic expect error

Commit:
fix: ensure rollback context in a case of error

Commit:
chore: update Cargo.lock

obycode committed:
chore: fix clippy warning about `Instant`

obycode committed:
Merge branch 'develop' into fix/clarity-wasm-context-rollback

obycode committed:
chore: additional uses of `time::Instant`

Commit:
chore: enable multi-threaded runtime for Tokio and require rusqlite

- Updated Tokio dependency to version 1.15 with the rt-multi-thread feature enabled.
- Removed  from rusqlite so that it is now a required dependency.

Commit:
chore: ensure type-checking call is executed

This commit guarantees that the type-checking call is performed as expected, preventing potential issues with type validation.

obycode committed:
Merge pull request #5912 from csgui/fix/clarity-wasm-context-rollback

Ensure context rollback in an error situation

BowTiedWoo committed:
fix: ensure memory is allocated for writing args

hugoclrd committed:
Merge branch 'develop' into chore/clarity-wasm-merge-develop

BowTiedWoo committed:
fix: get required bytes & grow memory before writing

BowTiedWoo committed:
fix: iterate directly on values for tuples

+ simplify total bytes calculation for lists

BowTiedWoo committed:
fix: check both name of the fields are identical

+ compare length of the maps to ensure no extra args

obycode committed:
Merge pull request #5954 from hugocaillard/chore/clarity-wasm-merge-develop

Chore/clarity wasm merge develop

hugoclrd committed:
fix: no default features compilation

hugoclrd committed:
fix: add clarity/vrf feature

obycode committed:
Merge pull request #5953 from BowTiedWoo/fix/calrity-wasm-contract-call-oom

`contract-call?` fix OOM

obycode committed:
Merge pull request #5957 from hugocaillard/fix/no-default-feature-compilation

Commit:
Merge branch 'develop' into clarity-wasm-develop-sync

Commit:
chore: update Cargo.lock

Commit:
chore: fmt

Commit:
fix: remove unused import

obycode committed:
Merge pull request #6004 from csgui/clarity-wasm-develop-sync

Clarity wasm develop sync

obycode committed:
chore: adjust code owners for this branch

This will allow the rules to be adjusted so that PRs to this branch only
require approvals from the Clarity Wasm team.

obycode committed:
Merge pull request #6063 from obycode/clarity-wasm-codeowners

chore: adjust code owners for this branch

wileyj committed:
Updating CODEOWNERS for wasm branch

wileyj committed:
Updating CODEOWNERS for wasm branch

wileyj committed:
Moving CODEOWNERS to .github, matching develop branch

wileyj committed:
Moving CODEOWNERS to .github, matching develop branch

Commit:
fix: propagate analysys error back to caller

Commit:
feat: always wasm

BowTiedWoo committed:
fix: add early return for invalid test cases

wileyj committed:
Merge pull request #6069 from wileyj/chore/clarity-wasm-codeowners

Updating CODEOWNERS for wasm branch

wileyj committed:
Merge branch 'feat/clarity-wasm-develop' into wasm-develop/always-wasm

csgui committed:
Merge branch 'feat/clarity-wasm-develop' into wasm-issue-653/propagation-error

BowTiedWoo committed:
Merge branch 'feat/clarity-wasm-develop' into feat/clarity-wasm-undefined-variable-fix

wileyj committed:
Merge pull request #6165 from csgui/wasm-develop/always-wasm

Always run wasm runtime

csgui committed:
Merge pull request #6169 from BowTiedWoo/feat/clarity-wasm-undefined-variable-fix

Skip Version-Incompatible Clarity Tests to Prevent UndefinedVariable Errors

csgui committed:
Merge pull request #6166 from csgui/wasm-issue-653/propagation-error

Propagate error

BowTiedWoo committed:
test: fix unchecked intermediary response tests

BowTiedWoo committed:
fix: format code

Commit:
test: skip incompatible wasm test

Commit:
test: ignore incompatible wasm test

Commit:
chore: add comment

Commit:
chore: add comment

BowTiedWoo committed:
test: fix intermediary response test

+ ignore irrelevant tests

BowTiedWoo committed:
fix: format code and add comments

csgui committed:
Merge pull request #6183 from csgui/wasm-issue-659/write-attempt-in-readonly

[clarity-wasm-tests] fix test failing with WriteAttemptedInReadOnlyMode

csgui committed:
Merge pull request #6182 from csgui/wasm-issue-662/unknown-function

[clarity-wasm-tests] fix test failing with UnknownFunction

BowTiedWoo committed:
test: keep original test & add ignore flag

BowTiedWoo committed:
test: remove useless flag & rename test

csgui committed:
Merge branch 'feat/clarity-wasm-develop' into test/clarity-wasm-fix-intermediary-response-tests

csgui committed:
Merge pull request #6189 from BowTiedWoo/test/clarity-wasm-fix-intermediary-response-tests

[clarity-wasm-tests] Fix UncheckedIntermediaryResponse Error in Tests

Commit:
test: fix Unchecked TypeError

Commit:
test: ignore incompatible wasm test

BowTiedWoo committed:
test: ignore test

csgui committed:
Merge pull request #6205 from csgui/wasm-issue-656/intermediary-responses

[clarity-wasm-tests] fix CheckErrors::UncheckedIntermediaryResponses

csgui committed:
Merge branch 'feat/clarity-wasm-develop' into test/clarity-wasm-type-value-err

csgui committed:
Merge pull request #6208 from BowTiedWoo/test/clarity-wasm-type-value-err

[clarity-wasm-tests] fix test failing with Unchecked(CheckErrors::TypeValueError(_, _)

Commit:
tests: fix CheckErrors::NoSuchContract error in tests

csgui committed:
Merge pull request #6198 from csgui/wasm-issue-652/no-such-contract

[clarity-wasm-tests] fix tests failing with CheckErrors::NoSuchContract error

csgui committed:
Merge branch 'feat/clarity-wasm-develop' into wasm-issue-677/unchecked-type-error

csgui committed:
Merge pull request #6204 from csgui/wasm-issue-677/unchecked-type-error

[clarity-wasm-tests] fix Unchecked(TypeError(UIntType, IntType))

Commit:
test: fix unknown trait ref

BowTiedWoo committed:
test: ignore incompatible wasm test

Commit:
tests: fix large contracts errors

csgui committed:
Merge pull request #6216 from csgui/wasm-issue-661/trait-reference-unknown

[clarity-wasm-tests] fix test failing with CheckError::TraitReferenceUnknown

Commit:
tests: fix large contracts errors

csgui committed:
Merge pull request #6218 from csgui/wasm-issue-664/large-contracts-tests-issue

[clarity-wasm-tests] fix large contracts test failure

BowTiedWoo committed:
test: fix trait_stack_depth test

Commit:
tests: fix vm::tests::traits issues

BowTiedWoo committed:
fix: change comment

Commit:
chore: fmt

csgui committed:
Merge pull request #6220 from csgui/wasm-issue-665/vm-tests-traits

[clarity-wasm-tests] fix vm::tests::traits

csgui committed:
Merge pull request #6217 from BowTiedWoo/wasm-issue-663/assertion-failed

[clarity-wasm-tests] fix test_cc_trait_stack_depth assertion failure

BowTiedWoo committed:
feat: add init_versioned_contract_with_db fn

BowTiedWoo committed:
test: fix pox_2_contract_caller_units test

Commit:
tests: fix vm::tests::contracts issues

csgui committed:
Merge pull request #6222 from BowTiedWoo/wasm-issue-668/no-such-contract-error

[clarity-wasm-tests] fix chainstate::stacks::boot::contract_tests

Commit:
tests: fix clarity_vm::tests::costs issues

csgui committed:
Merge pull request #6223 from csgui/wasm-issue-667/vm-tests-contracts

[clarity-wasm-tests] fix vm::tests::contracts issues

csgui committed:
Merge pull request #6224 from csgui/wasm-issue-669/clarity-vm-tests-costs

[clarity-wasm-tests] fix clarity_vm::tests::costs issues

Commit:
chore: update cargo.lock

csgui committed:
Merge pull request #6232 from csgui/update-wasm-dependency

update cargo.lock

BowTiedWoo committed:
test: fix test_samples test

BowTiedWoo committed:
fix: rename function and argument

BowTiedWoo committed:
fix: remove optional on analysis_db

BowTiedWoo committed:
fix: rename to analysis

csgui committed:
Merge pull request #6244 from BowTiedWoo/wasm-issue-673/assertion-failed

[clarity-wasm-tests] fix clarity_cli::test::test_samples

hugoclrd committed:
fix: build clarity-vm for wasm

csgui committed:
Merge pull request #6259 from hugocaillard/fix/build-clarit-vm-for-wasm

fix: build clarity vm for wasm

Commit:
chore: update clarity-wasm dependency

csgui committed:
Merge pull request #6226 from csgui/wasm-issue-666/stackerdb-set-signer-slots-issue

[clarity-wasm-tests] stackerdb-set-signer-slots issue

Commit:
fix rollback handling in case of a runtime error

This change needs to be reverted, as it adds an additional call to global_context.rollback(),
 which in turn disrupts the function call context.

Commit:
chore: update Cargo.lock

csgui committed:
Merge pull request #6277 from csgui/fix/function-context-rollback

[clarity-wasm] fix function context rollback

BowTiedWoo committed:
feat: add `call_function` arguments validation

BowTiedWoo committed:
Merge pull request #6302 from BowTiedWoo/fix/contract-call-args-validation

Add Clarity Arguments Validation on `call_function`

Acaccia committed:
fix: contract-call? linked function will now parse a TraitIdentifier for the trait

wileyj committed:
Merge pull request #6345 from Acaccia/fix/wasm-contract-call-fetch-trait

Clarity-wasm: fix for clarity-wasm issue 700

Commit:
Merge branch 'develop' into update

Commit:
chore: fmt

Commit:
code changes due to merge conflicts

csgui committed:
Merge pull request #6357 from csgui/update-wasm-develop

update clarity-wasm-develop branch

Commit:
chore: update wasm dependency

csgui committed:
Merge pull request #6364 from csgui/update-wasm-dependency

update Cargo.lock

BowTiedWoo committed:
fix: create callable argument

BowTiedWoo committed:
fix: format code

Commit:
Merge branch 'develop' into clarity-wasm-sync

Commit:
code adjustments due to merge

Commit:
tests: skip multiversion integration tests

Commit:
update Cargo.lock

csgui committed:
Merge pull request #6433 from csgui/clarity-wasm-sync

sync feat/clarity-wasm-develop branch

BowTiedWoo committed:
Merge branch 'feat/clarity-wasm-develop' into wasm-issue-675/fix-callable-argument

Commit:
chore: update Cargo.lock

csgui committed:
Merge pull request #6438 from csgui/wasm-develop-update

wasm-develop update dependencies

BowTiedWoo committed:
Merge branch 'feat/clarity-wasm-develop' into wasm-issue-675/fix-callable-argument

BowTiedWoo committed:
Merge pull request #6416 from BowTiedWoo/wasm-issue-675/fix-callable-argument

Fix `test_embedded_trait`

Commit:
fix: memory offsets

This PR fix memory offset for return values and change a flag to include the offset and length of the value.

csgui committed:
Merge pull request #6541 from csgui/fix/clarity-wasm-issue-719

fix memory offsets

Commit:
chore: adapt clarity-wasm integration to changes on develop branch

Commit:
Merge branch 'develop' into wip

Commit:
tests: fix

Commit:
chore: address review comments

Commit:
chore: update Cargo.lock

Commit:
Merge branch 'develop' into sync-wasm-develop

Commit:
Merge pull request #6610 from christian-stacks/sync-wasm-develop

sync wasm develop

bogdan-stacks committed:
test: ignore irelevant test

bogdan-stacks committed:
Merge pull request #6624 from bogdan-stacks/test/ignore-runtime-costs-test

Ignore runtime_limit test

bogdan-stacks committed:
add feature flag for specific errors

bogdan-stacks committed:
add feature flag for specific errors

bogdan-stacks committed:
replace find with contains

bogdan-stacks committed:
clean up code

bogdan-stacks committed:
construct expected err based on feature

bogdan-stacks committed:
Merge pull request #6726 from bogdan-stacks/wasm-issue-702/add-feature-flag

Fix wasm-issue#702 add feature flag

bogdan-stacks committed:
Merge pull request #6732 from bogdan-stacks/wasm-issue-675/add-feature-flag

Fix wasm-issue-675 by adding feature flag for specific errors

bogdan-stacks committed:
feat: add support for `stacks-block-time`

bogdan-stacks committed:
Merge pull request #6800 from bogdan-stacks/feat/clarity-wasm-stacks-block-time-support

Add Support for `stacks-block-time`

bogdan-stacks committed:
feat: code duplication for `current-contract` keyword

bogdan-stacks committed:
Merge pull request #6815 from bogdan-stacks/feat/clarity-wasm-current-contract-support

Code Duplication for `current-contract` Keyword

bogdan-stacks committed:
fix: code dup for `contract-hash` linker function

bogdan-stacks committed:
Merge pull request #6856 from bogdan-stacks/feat/support-contract-hash

Support `contract-hash?`

Acaccia-stacks committed:
feat: add a linked function for to-ascii with principal

melcar-stacks committed:
feat: change map methods to handle errors gracefully

For the cost compution of map's methods we need to compute the cost and
then throwback the error when an error occurs. This forced us to handle
errors better.

Acaccia-stacks committed:
Merge pull request #6860 from Acaccia-stacks/feat/to-ascii-linked-function

Clarity-wasm: linked function for `to-ascii?`

melcar-stacks committed:
chore: apply review comments. use reference to determine if there was an
error

Acaccia-stacks committed:
refactor: remove useless in_mem_offset in pass_argument

Acaccia-stacks committed:
fix: wrong computation for numbers of bytes in a string-utf8 passed argument

melcar-stacks committed:
chore: fmt

melcar-stacks committed:
revert: deleted link_contract_hash_fn by accident

melcar-stacks committed:
Merge pull request #6893 from melcar-stacks/feat/maps-methods-changes-for-cost

feat: change map methods to handle errors gracefully

Acaccia-stacks committed:
Merge pull request #6982 from Acaccia-stacks/fix/wasm-contract-call-oom

Clarity-wasm: fixes for `contract-call?` oom issues

federico-stacks commented:
@hugo-stacks I suppose this can be closed, right?
